### PR TITLE
[SPARK-34745][SQL] Unify overflow exception error message of integral types

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

Current, the overflow exception error messages of integral types are different.
For Byte/Short type, the message is "... caused overflow"
For Int/Long, the message is "int/long overflow" since Spark is calling the "*Exact"(e.g. addExact, negateExact) methods from java.lang.Math.

We should unify the error message by changing the message of Byte/Short as "tinyint/smallint overflow"

### Why are the changes needed?

Standardize exception messages in Spark

### Does this PR introduce _any_ user-facing change?

Yes, change the error message.

### How was this patch tested?

Unit tests.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/numerics.scala [hunk 1]
```diff
@@ -26,7 +26,7 @@ import org.apache.spark.sql.types.Decimal.DecimalIsConflicted
 private[sql] object ByteExactNumeric extends ByteIsIntegral with Ordering.ByteOrdering {
   private def checkOverflow(res: Int, x: Byte, y: Byte, op: String): Unit = {
     if (res > Byte.MaxValue || res < Byte.MinValue) {
-      throw new ArithmeticException(s"$x $op $y caused overflow.")
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/numerics.scala:29]: As an end user, I would prefer to see which particular values caused the issue. Unification is nice but it shouldn't make user experience worse, IMHO.

Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/numerics.scala:29]: Yes, I thought about this. 
On the other way, we can align the error message of int/log to the error message of byte/short, which is more user-friendly. Then basically we have to re-implement the "exact" methods in Spark.
cc @cloud-fan

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/numerics.scala:29]: It's a bit risky to implement the "exact" methods ourselves, as JDK may update them in future versions. I'd rather add a try-catch to change the error message.

Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/numerics.scala:29]: > It's a bit risky to implement the "exact" methods ourselves, as JDK may update them in future versions. I'd rather add a try-catch to change the error message.

So which one do you prefer? 
1. Changing the error message of byte/short overflow as simply: "tinyint/smallint overflow" (I check PostgreSQL and it  also simply show error messages like "ERROR: integer out of range")
2.Add try/catch in int/long arithmetic operations and throw a new exception with details.

cc @MaxGekk @cloud-fan @maropu

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/numerics.scala:29]: Actually, we have > 100 places where we use `*Exact` methods from the standard library. Such methods just throw the exception w/o any clues about inputs:
```java
    public static int multiplyExact(int x, int y) {
        long r = (long)x * (long)y;
        if ((int)r != r) {
            throw new ArithmeticException("integer overflow");
        }
        return (int)r;
    }
```
As the first step, we could align our `ArithmeticException` to the standard library (just a few places). After that we can think of how we can improve such kind of exceptions in all possible places in Spark, and give users more context.

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/numerics.scala:29]: > For Int/Long, the message is "int/long overflow" since Spark is calling the "*Exact"(e.g. addExact, negateExact) methods from java.lang.Math.

The unification itself looks nice and I think we should add try-catch in the int/log cases first.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala [hunk 2]
```diff
@@ -293,15 +293,6 @@ object QueryExecutionErrors {
...
   }
 
-  def unaryMinusCauseOverflowError(originValue: Short): ArithmeticException = {
-    new ArithmeticException(s"- $originValue caused overflow.")
-  }
-
-  def binaryArithmeticCauseOverflowError(
-      eval1: Short, symbol: String, eval2: Short): ArithmeticException = {
-    new ArithmeticException(s"$eval1 $symbol $eval2 caused overflow.")
-  }
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:303]: Why did you remove these methods from this file? How about merging these funcs into a unified one like this?
```
  def arithmeticCauseOverflowError(overflowCause: String): ArithmeticException = {
    new ArithmeticException(s"$overflowCause caused overflow.")
  }
```

Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:303]: > Why did you remove these methods from this file?

The PR is to align the error message of byte/short to int/long's. After the change,`binaryArithmeticCauseOverflowError` is not used in anywhere.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala [hunk 3]
```diff
@@ -61,7 +61,7 @@ case class UnaryMinus(
         s"""
            |$javaType $originValue = ($javaType)($eval);
            |if ($originValue == $javaBoxedType.MIN_VALUE) {
-           |  throw QueryExecutionErrors.unaryMinusCauseOverflowError($originValue);
+           |  throw new ArithmeticException("${dataType.simpleString} overflow");
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala:64]: shall we put this error message in `QueryExecutionErrors` like `unaryMinusCauseOverflowError`?
